### PR TITLE
HTTP/3: Add CompleteAsync tests and fix various bugs

### PR DESCRIPTION
**Pull request by JamesNK**
* Replicate HTTP/2 CompleteAsync unit tests for HTTP/3
* Fix writing wrong header for status code 500
* Fix various errors related to aborting

### CODE REVIEW

**Comment by JamesNK** [comment]
@jkotalik Rebased

**Comment by jkotalik** [comment]
Looks to be some test failures in this PR. Did my changes cause them to start failing?

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs [hunk 1]
```diff
@@ -377,8 +401,6 @@ private bool TryClose()
 
                 await Input.CompleteAsync();
 
-                await RequestBodyPipe.Writer.CompleteAsync();
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: Removed because errors weren't being correctly reported by request pipe writer. `CompleteAsync` was being called here before `CompleteAsync(Exception)` was being called somewhere else later.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs [hunk 2]
```diff
@@ -98,15 +98,39 @@ public Http3Stream(Http3Connection http3Connection, Http3StreamContext context)
...
-            Abort(ex, Http3ErrorCode.InternalError);
+            AbortCore(ex, Http3ErrorCode.InternalError);
         }
 
-        public void Abort(ConnectionAbortedException ex, Http3ErrorCode errorCode)
+        public void ResetAndAbort(ConnectionAbortedException ex, Http3ErrorCode errorCode)
+        {
+            // TODO - Keep track of whether stream has been aborted before?
+
+            Log.Http3StreamResetAbort(TraceIdentifier, errorCode, ex);
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:108]: Seems like the only difference here is that we are logging that we are sending a reset. Why can't these be combined with the normal Abort path with ResetAndAbort?

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs (hunk 1 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: Where is the call to CompleteAsync(Exception)? I imagine it's due to abort being called somehow?

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs [hunk 3]
```diff
@@ -760,5 +761,857 @@ public async Task ResetStream_ReturnStreamError()
 
             await requestStream.WaitForStreamErrorAsync(Http3ErrorCode.RequestCancelled, CoreStrings.FormatHttp3StreamResetByApplication(Http3ErrorCode.RequestCancelled));
         }
+
+        [Fact]
+        public async Task CompleteAsync_BeforeBodyStarted_SendsHeadersWithEndStream()
```
Comment (jkotalik) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs:766]: Are these testing the same functionality in HTTP/2 and 1.1?

Not going to thoroughly review these tests if that's the case.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs [hunk 4]
```diff
@@ -396,9 +397,11 @@ internal async Task SendDataAsync(Memory<byte> data, bool endStream = false)
             internal async Task<Dictionary<string, string>> ExpectHeadersAsync()
             {
                 var http3WithPayload = await ReceiveFrameAsync();
+                Assert.Equal(Http3FrameType.Headers, http3WithPayload.Type);
+
                 _decodedHeaders.Clear();
                 _qpackDecoder.Decode(http3WithPayload.PayloadSequence, this);
-                return _decodedHeaders;
+                return _decodedHeaders.ToDictionary(kvp => kvp.Key, kvp => kvp.Value, _decodedHeaders.Comparer);
```
Comment (jkotalik) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs:404]: Why was this needed?

Comment (JamesNK) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs:404]: Trailers will replace headers if a copy isn't made.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs (hunk 3 above)
Comment (JamesNK) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs:766]: Yes

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: `RequestBodyPipe.Writer.Complete(abortReason);` is called in the new `AbortCore`.

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: Is AbortCore called on each stream now?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: Do you mean from here?

https://github.com/dotnet/aspnetcore/blob/bb0be8c3148d2b28778eb313da6643cf4000f4ea/src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs#L353-L356

Yes

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: Hm, maybe that shouldn't be happening...

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs (hunk 4 above)
Comment (jkotalik) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs:404]: Ah I see.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: I don't know enough about this area to offer a good opinion, but I looked at `Http2Connection` and it calls into equivilent functionality on `Http2Stream`.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:380]: The biggest difference is Abort in HTTP/2 keeps track of whether it has already aborted - https://github.com/dotnet/aspnetcore/blob/bb0be8c3148d2b28778eb313da6643cf4000f4ea/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs#L532-L538

I left a TODO comment in Http3Stream.Abort about whether this needs to be tracked.

**Review by jkotalik — APPROVED** [review]
